### PR TITLE
RDoc-1529/1544 What is a Session and How Does it Work

### PR DESCRIPTION
**Pull request by morhilai**

### CODE REVIEW

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 1]
```diff
@@ -1,61 +1,97 @@
...
-- [Basics](../../indexes/querying/basics)
-
-### Document Store
-
-- [What is a Document Store](../../client-api/what-is-a-document-store)
+﻿# Session: What is a Session and How Does it Work
+---
+{NOTE: }
+
+The **Session** is:
```
Comment (Danielle9897) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Remove this line. 
Not all bullets bellow match it.

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 2]
```diff
@@ -1,61 +1,97 @@
...
+* Used to perform any database operation you might want. The basic CRUD operations are performed through `session`, more **advanced operations** are available through `session.advanced`.
+
+* Obtained from the `DocumentStore`.  
+  * Read more about [The Document Store](../../client-api/what-is-a-document-store) and [How to Create One](../../client-api/creating-document-store).  
+
+* Disposable. (Implements [IDisposable](https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=netframework-4.7.2))  
+  * Open every session in a `using` statement to ensure proper disposal.  
+
+* Tracks all changes made to entities and combines them into a single write operation sent upon the next call to `SaveChanges()`. This minimizes the total number of calls that go to the server over the network.  
+
```
Comment (Danielle9897) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Replace all above with:
=>

* The **Session**, which is obtained from the [Document Store](...), is a [Unit of Work](....) that represent a single business transaction on a particular database.

* Using the Session, perform needed operations on your documents.
  e.g. store a new document, modify an existing document, query your database, etc.
  Any such operation '_loads_' the documents entities to the Session for tracking.

* The Session **tracks all changes** made to all entities that it has either loaded or stored.
  You don’t need to manually track the changes to these entities and decide what needs to be saved and what doesn’t. 
  The Session will do it for you.

* To minimize the total number of calls to the server,  these changes are combined & persisted to the database only when calling `SaveChanges()`.

* Basic documents CRUD actions are available through the `Session`.
   For more advanced options,  use `Session.Advanced`.

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 3]
```diff
@@ -1,61 +1,97 @@
...
+      * [Batching](../../client-api/session/what-is-a-session-and-how-does-it-work#batching)  
+  * [Identity Map Pattern](../../client-api/session/what-is-a-session-and-how-does-it-work#identity-map-pattern)
+  * [Select N+1 Problem](../../client-api/session/what-is-a-session-and-how-does-it-work#select-n+1-problem)
+
+{NOTE/}  
+
+---
+{PANEL:Simple Example of Session Usage}
+
+The Client API, and the session in particular, is designed to be as straightforward as possible. This example code shows how we can use a session to store a document in the database.
```
Comment (Danielle9897) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Replace above with:
=> 

* The Client API, and the Session in particular, is designed to be as straight forward as possible.
  Open the session, do some operations, and apply the changes to the RavenDB server. 

* The example below shows how to store a new document in the database using the Session.

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 4]
```diff
@@ -1,61 +1,97 @@
...
+
+---
+{PANEL:Simple Example of Session Usage}
+
+The Client API, and the session in particular, is designed to be as straightforward as possible. This example code shows how we can use a session to store a document in the database.  
+
+{CODE session_usage_1@ClientApi\Session\WhatIsSession.cs /}
+
+Calling `SaveChanges()` sends the entity to the RavenDB server where it is stored as a new document in the database. (Since no other database was specified, 
+this will be the [Default Database](http://localhost:54391/docs/article-page/4.1/csharp/client-api/setting-up-default-database)).
```
Comment (Danielle9897) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Replace above 2 lines with:
=>

* The `SaveChanges()` call checks the Session state for all changes made to the tracked entities.
  These changes are sent to the server as a single remote call that will complete transactionally.
  In other words, either all changes are saved as a single unit or none of them are.
  Once `SaveChanges()` returns, it is guaranteed that the changes are persisted to the database.

* The `SaveChanges()` is the only time when a RavenDB client sends updates to the server, so you will experience a reduced number of network calls.

* In the above example,  Calling `SaveChanges()` actually sends the new entity to the RavenDB server where it is persisted as a new document in the database.  Since no other database was specified when opening the Session, the [Default Database](...) is used.

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 5]
```diff
@@ -1,63 +1,60 @@
...
+    {
+        public WhatIsSession()
+        {
+            using (var store = new DocumentStore())
+            {
+                // A client-side copy of the document Id
+                string companyId = null;
+
+                #region session_usage_1
+                // Store new document
```
Comment (Danielle9897) [inline Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Change to:
=>
// Obtain a Session from your Document Store

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 6]
```diff
@@ -1,63 +1,60 @@
...
+                using (IDocumentSession session = store.OpenSession())
+                {
+                    Company entity = new Company { Name = "Company" };
```
Comment (Danielle9897) [inline Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Add comment above this line
=>
// Create a new entity

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 7]
```diff
@@ -1,63 +1,60 @@
...
+                    session.Store(entity);
```
Comment (Danielle9897) [inline Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Add comment above this line
=>
// Mark the entity for storage in the Session

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 8]
```diff
@@ -1,63 +1,60 @@
...
+                    // The entity is now marked for storage.
+                    // Changes made to it will be tracked by default.
+                    // However it is not actually sent to the server 
+                    // until `SaveChanges()` is called.
```
Comment (Danielle9897) [inline Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Replace above comments with:
=>
// Any changes now made to the entity will be tracked by the Session
// However, the changes are persisted to the server only when 'SaveChanges()' is called.

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 9]
```diff
@@ -1,61 +1,97 @@
...
+
+* Tracks all changes made to entities and combines them into a single write operation sent upon the next call to `SaveChanges()`. This minimizes the total number of calls that go to the server over the network.  
+  
+
+* In this page:  
+  * [Simple Example of Session Usage](../../client-api/session/what-is-a-session-and-how-does-it-work#simple-example-of-session-usage)  
+  * [Unit of Work Pattern](../../client-api/session/what-is-a-session-and-how-does-it-work#unit-of-work-pattern)  
+      * [Batching](../../client-api/session/what-is-a-session-and-how-does-it-work#batching)  
+  * [Identity Map Pattern](../../client-api/session/what-is-a-session-and-how-does-it-work#identity-map-pattern)
+  * [Select N+1 Problem](../../client-api/session/what-is-a-session-and-how-does-it-work#select-n+1-problem)
```
Comment (Danielle9897) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Change structure to:
=> 

* In this page:

  * Unit of Work Pattern
       * Tracking Changes
       * Batching 
       * Store New Document Example
       * Load & Edit Document Example

  *  Identity Map Pattern

  * Reducing Server Calls For:
      * The N+1 Problem
      * Large query results
      * Retrieving results on demand (Lazy)

(Note: _Reducing Server Calls_ may deserve a dedicated page of its own)

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 10]
```diff
@@ -1,61 +1,97 @@
...
+
+Should it arise, RavenDB offers a number of ways to mitigate this problem:  
+
+* [Perform Queries Lazily](https://ravendb.net/docs/article-page/4.1/csharp/client-api/session/querying/how-to-perform-queries-lazily)  
+* [Stream Query Results](https://ravendb.net/docs/article-page/4.1/csharp/client-api/session/querying/how-to-stream-query-results)  
+
+{INFO: Configuring Maximum Requests per Session} 
+By default, the maximum number of requests allowed per session is 30. Exceeding this limit causes an exception to be thrown.
+This limit can be changed at the `DocumentConventions::MaxNumberOfRequestsPerSession` property.
+{INFO/}
```
Comment (Danielle9897) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Remove above 'INFO' form here.
It is now mentioned in the bullets in the overview.

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 11]
```diff
@@ -1,61 +1,97 @@
...
+Remote calls to a server are among the most expensive operations your application can make. The session optimizes this by batching all write operations it has tracked 
+into a single call to the server, sent when `SaveChanges()` is called. This is the default behavior of the session; it doesn't need to be enabled. This also ensures that all 
+writes made to the database in a single session are executed as a single atomic transaction, no matter how many operations you are actually executing.  
+
+{INFO:How to Disable Entity Tracking}
+Entity tracking can be disabled using the `SessionOptions.NoTracking` property when a session is being 
+opened. [Read more](../../client-api/session/opening-a-session#example-ii---disabling-entities-tracking).  
+{INFO/}
+
+{PANEL/}
```
Comment (Danielle9897) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Replace all above Panels (line 30 - line 58) with the following:

## For PANEL  **'Unit of Work Pattern'**:

  ###  Tracking Changes   
   
 *  Using the Session, perform needed operations on your documents.
    e.g. store a new document, modify an existing document, query your database, etc.
    Any such operation '_loads_' the documents entities to the Session for tracking.

 * The Session **tracks all changes** made to all entities that it has either loaded or stored.
  You don’t need to manually track the changes to these entities and decide what needs to be saved and what doesn’t.  
 The Session will do it for you.

* All these tracked changes are combined & persisted to the database only when calling `SaveChanges()`.

* Entity tracking can be disabled if needed. See [Disabling Entities Tracking](...)
  (use link ../../..client-api/session/opening-a-session#example-ii---disabling-entities-tracking)

###  Batching

* Remote calls to a server over the network are among the most expensive operations an application makes.
  The session optimizes this by batching all write operations it has tracked into the single SaveChanges() call.

* The SaveChanges() call checks the Session state for all changes made to the tracked entities.
 These changes are sent to the server as a single remote call that will complete transactionally.
 In other words, either all changes are saved as a **Single Atomic Transaction** or none of them are.
 Once SaveChanges() returns, it is guaranteed that the changes are persisted to the database.
  
* The SaveChanges() is the only time when a RavenDB client sends updates to the server, so you will experience a 
 reduced number of network calls.

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 12]
```diff
@@ -1,63 +1,60 @@
...
+                    session.SaveChanges();
```
Comment (Danielle9897) [inline Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Add comment below SaveChanges()
=>
// At this point the entity is persisted to the database as a new document.
// Since no database was specified when opening the Session, the Default Database is used.

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 13]
```diff
@@ -1,63 +1,60 @@
...
+                }
+                #endregion
+
+                #region session_usage_2
+                // Load document from database, edit it, and save it
```
Comment (Danielle9897) [inline Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Change above comment to:
=>
// Open a session

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 14]
```diff
@@ -1,63 +1,60 @@
...
+                using (IDocumentSession session = store.OpenSession())
+                {
+                    // Load using known Id
```
Comment (Danielle9897) [inline Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Change above comment to:
=>
// Load an existing document to the Session using its Id

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 15]
```diff
@@ -1,63 +1,60 @@
...
+                    Company entity = session.Load<Company>(companyId);
+                    // Edit a property
```
Comment (Danielle9897) [inline Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Change above comment to:
=>
// Edit the entity. The Session will track this change.

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 16]
```diff
@@ -1,63 +1,60 @@
...
+                    entity.Name = "Another Company";
+
+                    // When a document is loaded, the session tracks 
+                    // all changes made to it by default.
+                    // A call to `SaveChanges()` sends all accumulated changes to the server.
+                    session.SaveChanges(); 
+                    // Note that no call to `Store()` is required for a loaded document.
```
Comment (Danielle9897) [inline Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Replace above comment with:
=>
// At this point, the change made is persisted in the database.

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 17]
```diff
@@ -1,61 +1,97 @@
...
+
+{PANEL:Identity Map Pattern}
+
+The session implements the [Identity Map Pattern](https://martinfowler.com/eaaCatalog/identityMap.html). This means that in the context of a single session, the same document loaded twice 
+always resolves to the same entity instance. The first call to `Load()` will retrieve a document from the database and save it in the identity map. The second call to `Load()` 
+will simply retrieve the same document from the identity map and will not make an additional call to the server.
```
Comment (Danielle9897) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Replace above with the following bullets:
=>

* The session implements the [Identity Map Pattern](..).

* The first document `Load()` call goes to the server and fetches the document from the database.
  The document is then saved as an entity in the Session's identity map.

* All subsequent `Load()` calls to the same document will simply retrieve the entity from the Session - no additional calls to the server are made. 

* Note: To override this and update an entity with the latest changes from the server see: [Refresh an Entity](...)
  (use link: ../../../client-api/session/how-to/refresh-entity)

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 18]
```diff
@@ -1,61 +1,97 @@
...
+{PANEL/}
+
+{PANEL:Identity Map Pattern}
+
+The session implements the [Identity Map Pattern](https://martinfowler.com/eaaCatalog/identityMap.html). This means that in the context of a single session, the same document loaded twice 
+always resolves to the same entity instance. The first call to `Load()` will retrieve a document from the database and save it in the identity map. The second call to `Load()` 
+will simply retrieve the same document from the identity map and will not make an additional call to the server. 
+{CODE session_usage_3@ClientApi\Session\WhatIsSession.cs /}
+
+The above command will not throw an exception.
```
Comment (Danielle9897) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Remove above line from here. 
Insert as a comment in the code.

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 19]
```diff
@@ -1,63 +1,60 @@
...
+                }
+                #endregion
+
+                using (var session = store.OpenSession())
+                {
+                    #region session_usage_3
+                    Assert.Same(session.Load<Company>(companyId), session.Load<Company>(companyId));
```
Comment (Danielle9897) [inline Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Add the comment about 'no exception' here

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 20]
```diff
@@ -1,61 +1,97 @@
...
+
+The [Select N+1 problem](http://blogs.microsoft.co.il/gilf/2010/08/18/select-n1-problem-how-to-decrease-your-orm-performance/) is very common with all ORMs and ORM-like APIs, including the RavenDB Client API. It results in an excessive number of remote calls to the server, which makes a query very expensive.
+
+The select n+1 problem should never arise if RavenDB is being used correctly. The number of remote calls per session should be as close to 1 as possible. If the maximum requests limit is reached, it is a sure sign of either select n+1 or some other misuse of the session.
+
+Should it arise, RavenDB offers a number of ways to mitigate this problem:  
+
+* [Perform Queries Lazily](https://ravendb.net/docs/article-page/4.1/csharp/client-api/session/querying/how-to-perform-queries-lazily)  
+* [Stream Query Results](https://ravendb.net/docs/article-page/4.1/csharp/client-api/session/querying/how-to-stream-query-results)  
+
```
Comment (Danielle9897) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Replace above Panel with:
=>

## PANEL Reducing Server Calls For:

###  The N+1 Problem
   * The [Select N+1 problem](...) is common with all ORMs and ORM-like APIs.
     It results in an excessive number of remote calls to the server, which makes a query very expensive.session.

   * Make use of RavenDB's `include()` method to include related documents and avoid this issue. 
     See: [Document Relationships](....)
     (use link ../../../client-api/how-to/handle-document-relationships)

### Large query results
   * When query results are large and you don't want the overhead of keeping all results in memory,
      you can [Stream the Query Results](...). 
      A single server call is executed and the client can handle the results one by one.

  * Note: [Paging](...) also avoids getting all query results in one time, but multiple server calls are generated - one per page retrieved.

### Retrieving results on demand (Lazy)
 * Query calls to the server can be delayed and executed on-demand as needed using `Laxily()`
*  See [Perform Queries Lazily](...)

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown (hunk 11 above)
Comment (morhilai) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: In this sentence:
"Any such operation 'loads' the documents entities to the Session for tracking."
Did you mean to use both "documents" and "entities"?

Comment (Danielle9897) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Can use   ....'loads' the documents to the Session ... to keep it simple

Comment (Danielle9897) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: image

Comment (Danielle9897) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: image

Comment (Danielle9897) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: image

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown (hunk 20 above)
Comment (Danielle9897) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: image

Comment (morhilai) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Done

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown (hunk 1 above)
Comment (morhilai) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Done

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown (hunk 2 above)
Comment (morhilai) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs (hunk 5 above)
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs (hunk 6 above)
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs (hunk 7 above)
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs (hunk 8 above)
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown (hunk 9 above)
Comment (morhilai) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Done

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown (hunk 10 above)
Comment (morhilai) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Done

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown (hunk 11 above)
Comment (morhilai) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs (hunk 12 above)
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs (hunk 13 above)
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs (hunk 14 above)
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs (hunk 15 above)
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs [hunk 21]
```diff
@@ -1,63 +1,60 @@
...
+                using (IDocumentSession session = store.OpenSession())
+                {
+                    // Load using known Id
+                    Company entity = session.Load<Company>(companyId);
+                    // Edit a property
+                    entity.Name = "Another Company";
+
+                    // When a document is loaded, the session tracks 
+                    // all changes made to it by default.
+                    // A call to `SaveChanges()` sends all accumulated changes to the server.
```
[the remark this answers is not in the document]
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs (hunk 16 above)
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown (hunk 17 above)
Comment (morhilai) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Done

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown (hunk 18 above)
Comment (morhilai) [reply Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: Done

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs (hunk 19 above)
Comment (morhilai) [reply Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/WhatIsSession.cs]: Done

File: Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown [hunk 22]
```diff
@@ -1,61 +1,120 @@
...
+#### Tracking Changes
+* Using the Session, perform needed operations on your documents.  
+  e.g. store a new document, modify an existing document, query your database, etc.  
+  Any such operation '*loads*' the documents to the Session for tracking.  
+* The Session **tracks all changes** made to all entities that it has either loaded or stored.  
+  You don’t need to manually track the changes to these entities and decide what needs to be saved and what doesn’t.  
+  The Session will do it for you.  
+* All these tracked changes are combined & persisted to the database only when calling `SaveChanges()`.  
+* Entity tracking can be disabled if needed. See:  
+  * [Disabling Entity Tracking](../../client-api/session/opening-a-session#example-ii---disabling-entity-tracking)
```
Comment (morhilai) [inline Documentation/4.1/Raven.Documentation.Pages/client-api/session/what-is-a-session-and-how-does-it-work.dotnet.markdown]: This link will be dead once the changes to "Session: Opening a Session" are uploaded. Changing to:
=> * [Disable Entity Tracking](../../client-api/session/configuration/how-to-disable-tracking)
